core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,213 @@
+  // }
+
+  // /**
+  //  * Maps from `EA` error type to another.
+  //  */
+  // final def mapErrorSet[E](f: EA => E): ZRef[E, EB, A, B] = new ZRef[E, EB, A, B] {
+  //   def set(a: A): IO[E, Unit]                                              = self.set(a).mapError(f)
+  //   def get: IO[EB, B]                                                      = self.get
+  //   def setAsync(a: A): zio.IO[E, Unit]                                     = self.setAsync(a).mapError(f)
+  //   private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, B]           = self.unsafeUpdate(f)
+  //   private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): zio.IO[E, C] = self.unsafeModify(f)
+  // }
+
+  // /**
+  //  * Create a new Ref unifying the Error types `EA` and `EB` to one `E`
+  //  * via the conversion functions `ea` and `eb` .
+  //  */
+  // final def unifyError[E](ea: EA => E, eb: EB => E): ZRef[E, E, A, B] =
+  //   new ZRef[E, E, A, B] {
+  //     def set(a: A): IO[E, Unit]                             = self.set(a).mapError(ea)
+  //     def get: IO[E, B]                                      = self.get.mapError(eb)
+  //     def setAsync(a: A): IO[E, Unit]                        = self.setAsync(a).mapError(ea)
+  //     def unsafeUpdate[E](f: Any => Any): IO[E, B]           = self.unsafeUpdate(f)
+  //     def unsafeModify[E, C](f: Any => (Any, Any)): IO[E, C] = self.unsafeModify(f)
+  //   }
+
+  // /**
+  //  * Create a new Ref unifying the values `A` and `B` to one `C`
+  //  * via the conversion functions `ca` and `bc` .
+  //  */
+  // final def unifyValue[C](ca: C => A, bc: B => C): ZRef[EA, EB, C, C] =
+  //   new ZRef[EA, EB, C, C] {
+  //     def set(c: C): IO[EA, Unit]                            = self.set(ca(c))
+  //     def get: IO[EB, C]                                     = self.get.map(bc)
+  //     def setAsync(c: C): IO[EA, Unit]                       = self.setAsync(ca(c))
+  //     def unsafeUpdate[E](f: Any => Any): IO[E, C]           = self.unsafeUpdate(f).map(bc)
+  //     def unsafeModify[E, C](f: Any => (Any, Any)): IO[E, C] = self.unsafeModify(f)
+  //   }
+
+  // /**
+  //  * Discard the content of the Ref and replace it with `Unit`
+  //  */
+  // final def unit: ZRef[EA, EB, A, Unit] = const(())
+
+  // /**
+  //  *  used for performance reasons and go around variance issues
+  //  */
+  // private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): IO[E, C]
+
+  // /**
+  //  * used for performance reasons and go around variance issues
+  //  */
+  // private[zio] def unsafeUpdate[E](f: Any => Any): IO[E, B]
+
+  def update(f: B => A): IO[E, B] =
+    get.flatMap(b => set(f(b)) *> get).uninterruptible
+
+  def modify[C](f: B => (C, A)): IO[E, C] =
+    get.flatMap { b =>
+      val (c, a) = f(b)
+      set(a).map(_ => c)
+    }.uninterruptible
+
+}
+object ZRef extends Serializable {
+  implicit class InvariantSyntax[E, A](ref: ZRef[E, A, A]) {
+    def updateSome(pf: PartialFunction[A, A]): IO[E, A] =
```

Comment:
Can these move back into `ZRef` to? As
```
def updateSome(pf: PartialFunction[B, A]): IO[E, A]
def modifySome[C](default: C)(pf: PartialFunction[B, (C, A)])
```

Reply:
Compiler complains for 
for `def updateSome(pf: PartialFunction[B, A]): IO[E, A]`
`contravariant type A occurs in covariant position in type (pf: PartialFunction[B,A])zio.IO[E,A] of method updateSome`

For `modifySome` I am stuck here, I need an `A` : 
`def modifySome[C](default: C)(pf: PartialFunction[B, (C, A)]) = modify(pf.orElse[B, (C, A)] { case b => (default, ???) })`

Reply:
My bad, it should be:
```scala
def updateSome(pf: PartialFunction[B, A]): IO[E, B]
def modifySome[C](default: C)(pf: PartialFunction[B, (C, A)]): IO[E, B]
```
just like the `update` and `modify` equivalents.

Reply:
There might be a formulation that does not use `orElse` so you don't fall into that problem.